jni/lib/src/jni_object.dart

```diff
@@ -7,178 +7,543 @@ import 'dart:ffi';
+    _env.GetFloatField,
+    _env.GetStaticFloatField,
+    _env.CallFloatMethodA,
+    _env.CallStaticFloatMethodA,
+  ),
+  JniType.doubleType: _CallGetMethods(
+    _env.GetDoubleField,
+    _env.GetStaticDoubleField,
+    _env.CallDoubleMethodA,
+    _env.CallStaticDoubleMethodA,
+  ),
+  JniType.objectType: _CallGetMethods(
+    _env.GetObjectField,
+    _env.GetStaticObjectField,
+    _env.CallObjectMethodA,
+    _env.CallStaticObjectMethodA,
+  ),
+  JniType.voidType: _CallGetMethods(
+    (x, y) => throw ArgumentError('void passed as field type'),
+    (x, y) => throw ArgumentError('void passed as static field type'),
+    _env.CallVoidMethodA,
+    _env.CallStaticVoidMethodA,
+  ),
+};
+
+T _getID<T>(
+    T Function(Pointer<Void> ptr, Pointer<Char> name, Pointer<Char> sig) f,
+    Pointer<Void> ptr,
+    String name,
+    String sig) {
+  final result = using(
+      (arena) => f(ptr, name.toNativeChars(arena), sig.toNativeChars(arena)));
+  _env.checkException();
+  return result;
+}
+
+int _getCallType(int? callType, int defaultType, Set<int> allowed) {
+  if (callType == null) return defaultType;
+  if (allowed.contains(callType)) return callType;
+  throw InvalidCallTypeException(callType, allowed);
+}
+
+T _callOrGet<T>(int? callType, Function(int) f) {
+  final int finalCallType;
+  T result;
+  switch (T) {
+    case bool:
+      finalCallType =
+          _getCallType(callType, JniType.boolType, {JniType.boolType});
+      result = (f(finalCallType) as int != 0) as T;
+      break;
+    case int:
+      finalCallType = _getCallType(callType, JniType.intType, {
+        JniType.byteType,
+        JniType.charType,
+        JniType.shortType,
+        JniType.intType,
+        JniType.longType,
+      });
+      result = f(finalCallType) as T;
+      break;
+    case double:
+      finalCallType = _getCallType(callType, JniType.doubleType,
+          {JniType.floatType, JniType.doubleType});
+      result = f(finalCallType) as T;
+      break;
+    case String:
+    case JniObject:
+    case JniString:
+      finalCallType =
+          _getCallType(callType, JniType.objectType, {JniType.objectType});
+      final ref = f(finalCallType) as JObject;
+      if (ref == nullptr) {
+        _env.checkException();
+      }
+      final ctor = T == String
+          ? (ref) => _env.asDartString(ref, deleteOriginal: true)
+          : (T == JniObject ? JniObject.fromRef : JniString.fromRef);
+      result = ctor(ref) as T;
+      break;
+    case _VoidType:
+      finalCallType =
+          _getCallType(callType, JniType.voidType, {JniType.voidType});
+      f(finalCallType);
+      result = null as T;
+      break;
+    case dynamic:
+      result = f(callType ?? JniType.voidType) as T;
+      break;
+    default:
+      throw UnsupportedError('Unknown type $T');
+  }
+  return result;
+}
+
+T _callMethod<T>(
+    int? callType, List<dynamic> args, Function(int, Pointer<JValue>) f) {
+  final jArgs = JValueArgs(args);
```

Comment:
Use Arena instead.

If `_callOrGet` can throw (or we change it in the future to throw) then the `free` two lines below doesn't happen).